storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -641,7 +642,12 @@ void cleanup() {
 
             try {
                 forEachTopologyDistDir((p, topologyId) -> {
-                    if (!safeTopologyIds.contains(topologyId)) {
+                    String topoJarKey = ConfigUtils.masterStormJarKey(topologyId);
+                    String topoCodeKey = ConfigUtils.masterStormCodeKey(topologyId);
+                    String topoConfKey = ConfigUtils.masterStormConfKey(topologyId);
+                    if (!topologyBlobs.containsKey(topoJarKey)
```

Comment:
It seems like it may be possible to still have a race condition here where a topology is added to topologyBlobs?

If so it seems like we should add synchronization to this code and the adding of a blob?

Reply:
My understanding is topo dir is always created after the entries of `topologyBlobs` are added. So whenever the topology entry is found, we should not delete the dir. 
Also, `topologyBlobs` is `ConcurrentHashMap` which allows the slot and cleanup/update threads to access in parallel. 
If we added synchronization, the file system dir cleanup might take seconds and block the slot threads to progress.

Reply:
What happens if line 648 executes and before line 649 runs getTopoJar(topologyId) is called?

Reply:
When that happens, the directory will not be deleted. It is okay because when `getTopoJar` executes, it means the directory is needed. 

The bottom line here is the code will delete the directory for dead topologies eventually (cleanup thread runs every 30s). But if there is anywhere that needs this directory, this code will not delete the directory

Reply:
@agresch Technically there is a chance. But with this change it would be more unlikely to happen than before.
Before this change, we compare topo dir to `safeTopologyIds` which is essentially a much older version of `topologyBlobs  snapshot since it could take 10s for the `forEachTopologyDistDir` loop to finish. In this case, race condition is much more likely to happen due to the gap. 
Now we directly check the `topologyBlobs` map right before deletion. So the `forEachTopologyDistDir` loop could be aware of the concurrent additions from slots and further prevent most of the wrongful deletion to happen.

Reply:
I guess your question is if line 650 executes and before line 651 runs, getTopoJar(topologyId) is called?
Yes there is a chance. But current change is much better than what we have before.


Just for future reference, one thing about Rui's comment that I want to note is that we still don't know where the 10s delay we observed from the incident we had in our production cluster between where `safeTopologyIds` is populated and `fsOps.deleteIfExists` is invoked comes from. It could come from `forEachTopologyDistDir` file operations. But we don't know yet. Also this issue doesn't alway happen.